src/transform/src/notices/index_key_empty.rs

```diff
@@ -0,0 +1,62 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Hosts [`IndexKeyEmpty`].
+
+use std::collections::BTreeSet;
+use std::fmt;
+
+use mz_repr::explain::ExprHumanizer;
+use mz_repr::GlobalId;
+
+use crate::notices::{ActionKind, OptimizerNoticeApi};
+
+/// An index with an empty key is maximally skewed (all of the data goes to a single worker),
+/// and is almost never really useful. It's slightly useful for a cross join, because a cross
+/// join also has an empty key, so we avoid rearranging the input. However, this is still
+/// not very useful, because
+///  - Rearranging the input shouldn't take too much memory, because if a cross join has a big
+///    input, then we have a serious problem anyway.
+///  - Even with the arrangement already there, the cross join will read every input record, so
+///    the orders of magnitude performance improvements that can happen with other joins when an
+///    input arrangement exists can't happen with a cross join.
+/// Also note that skew is hard to debug, so it's good to avoid this problem in the first place.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct IndexKeyEmpty;
+
+impl OptimizerNoticeApi for IndexKeyEmpty {
+    fn dependent_items(&self) -> BTreeSet<GlobalId> {
+        BTreeSet::new()
```

Comment:
Why is this empty, as opposed to containing the index ID and the relation ID that the index is on?

Reply:
Currently there is no way to pass the `id` for the enclosing index while we are optimizing the index because we don't have yet wired this information to our transforms. When the index or the backing `on` object are deleted we will also delete the notice because we are going to hit this loop in `Catalog::transact`

```rust
for id in drop_ids {
    if let Some(df_meta) = self.try_get_dataflow_metainfo(&id) {
        // Generate retractions for the Builtin tables.
        self.pack_optimizer_notices(
            &mut builtin_table_updates,
            df_meta.optimizer_notices.iter(),
            -1,
        );
    }
    // Drop in-memory planning metadata.
    self.drop_plans_and_metainfos(id);
}
```